### PR TITLE
Read SourceLink info and call service to retrieve source from there

### PR DESCRIPTION
**Pull request by davidwengier**
Part of https://github.com/dotnet/roslyn/issues/55834

This creates a source link service and uses it to download files, but there is no source link service implementation in here. That will come in a separate PR to main-vs-deps because it needs new debugger contracts. There is also a new logger interface that will be filled in later, in another PR, but better to plumb it through now so that the main-vs-deps change can be singular in nature

This also changes things so we only write a temp file for embedded source, and just use the existing source file if possible, or the source link source file, which the debugger will write to its temp folder. The latter is especially important so that breakpoints work.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/PdbSourceDocument/AbstractPdbSourceDocumentTests.cs
Comment (davidwengier) [inline src/EditorFeatures/CSharpTest/PdbSourceDocument/AbstractPdbSourceDocumentTests.cs:1]: All of this is just a move from `PdbSourceDocumentTests`

File: src/EditorFeatures/CSharpTest/PdbSourceDocument/NullResultMetadataAsSourceFileProvider.cs
Comment (davidwengier) [inline src/EditorFeatures/CSharpTest/PdbSourceDocument/NullResultMetadataAsSourceFileProvider.cs:1]: This is a just move and rename

File: src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj
Comment (davidwengier) [inline src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj:133]: This is a source package, and I've verified F5ing with VS Preview works totally fine, so I think this is okay to target main?

Comment (davidwengier) [reply src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj:133]: Anyone know how to ignore formatting warnings/errors from this? Adding `NoWarn` to the package reference didn't work

File: src/Features/Core/Portable/GlobalSuppressions.cs
Comment (davidwengier) [inline src/Features/Core/Portable/GlobalSuppressions.cs:8]: IDE0055 suppression here works fine, but IDE0073 doesn't, which is not surprising since its not tied to a type :(

File: src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj
Comment (ryzngard) [reply src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj:133]: Not sure if we'll need a validation build or handholding for new dependencies through RPS

File: src/Features/Core/Portable/PdbSourceDocument/DocumentDebugInfoReader.cs
Comment (ryzngard) [inline src/Features/Core/Portable/PdbSourceDocument/DocumentDebugInfoReader.cs:65]: It's been a bit since I looked at the contract, but does _pdbReader throw if you provide a name that `isNil`?

File: src/Features/Core/Portable/PdbSourceDocument/IPdbSourceDocumentLogger.cs
Comment (ryzngard) [inline src/Features/Core/Portable/PdbSourceDocument/IPdbSourceDocumentLogger.cs:11]: It's not clear to me why this was added... why not just `ILogger`?

File: src/Features/Core/Portable/PdbSourceDocument/PdbFileLocatorService.cs
Comment (ryzngard) [inline src/Features/Core/Portable/PdbSourceDocument/PdbFileLocatorService.cs:58]: clever

File: src/Features/Core/Portable/PdbSourceDocument/PdbFileLocatorService.cs
Comment (ryzngard) [inline src/Features/Core/Portable/PdbSourceDocument/PdbFileLocatorService.cs:69]: When can this happen?

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs
Comment (ryzngard) [inline src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:50]: interesting, is there really much gain from keeping the file on disk instead of pulling it out fresh every time? Seems like unnecessary complexity

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs
Comment (ryzngard) [inline src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:117]: Not a huge deal, but I could see this being useful in [TaskExtensions](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/TaskExtensions.cs,599b9667820e1f18)

Comment (ryzngard) [reply src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:117]: Something like, `Task.RunWithTimeout<T>(Task<T> taskToRun, int timeoutInMs, CancellationToken cancellationToken)`

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs
Comment (ryzngard) [inline src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs:129]: ❓Is there any specific reason this doesn't just use `IOUtilities.PerformIO`?

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs
Comment (ryzngard) [inline src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs:148]: is this needed? I would expect it to know it's not nullable based on the `LINQ` calls above

File: src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj
Comment (ryzngard) [inline src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj:138]: I don't understand how these are part of this change

**Review by ryzngard — APPROVED** [review]
This looks as good as I could guess for the change. Some open questions that I'd like answered but not worth blocking the PR for.

File: src/Features/Core/Portable/PdbSourceDocument/ISourceLinkService.cs
Comment (davidwengier) [reply src/Features/Core/Portable/PdbSourceDocument/ISourceLinkService.cs:18]: So this is just a direct remapping of the corresponding type from MS.VS.Debugger.Contracts, but we can't reference that in the features layer. Which is to say what "Log" is might make you sad, but I can certainly document it :)

File: src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj
Comment (davidwengier) [reply src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj:138]: I'm not entirely sure either, but the story I'm telling myself is that System.Text.Json is a transitive dependencies from the SourceLink.Tools package, and Text.Encodings.Web is transitive of that, and I'm the lucky winner of being the first to reference System.Text.Json in Roslyn.

Basically I added the package reference, built on the command line, and an error told me to modify this file :)

File: src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj
Comment (davidwengier) [reply src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj:133]: A source package means that adding this reference just copies https://github.com/dotnet/sourcelink/blob/main/src/SourceLink.Tools/SourceLinkMap.cs into the obj directory, and adds it to the project, so shouldn't be any new DLL loads or anything.

File: src/Features/Core/Portable/PdbSourceDocument/DocumentDebugInfoReader.cs
Comment (davidwengier) [reply src/Features/Core/Portable/PdbSourceDocument/DocumentDebugInfoReader.cs:65]: I don't know, but I'm checking it just above. Or were you suggesting the check is redundant? I'm a belt and braces kinda person I guess (except when it comes to network and file IO which I seem to always forget about)

File: src/Features/Core/Portable/PdbSourceDocument/IPdbSourceDocumentLogger.cs
Comment (davidwengier) [reply src/Features/Core/Portable/PdbSourceDocument/IPdbSourceDocumentLogger.cs:11]: These logs are specifically for trying to help library authors, if source link isn't configured properly or something, and their symbols aren't right or something. They go to the output window (screenshots in https://github.com/dotnet/roslyn/pull/57996). Our normal logs are visible easily, and we wouldn't want all of the extra noise from all of the other things that ILogger gets told about

File: src/Features/Core/Portable/PdbSourceDocument/PdbFileLocatorService.cs
Comment (davidwengier) [reply src/Features/Core/Portable/PdbSourceDocument/PdbFileLocatorService.cs:69]: Presumably never.

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs
Comment (davidwengier) [reply src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:50]: Probably not, but I don't think we save much by adding the complexity to figure out when to delete it either.

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs
Comment (davidwengier) [reply src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs:129]: No, good call. This was just lazily copied from the decompilation source provider

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs
Comment (davidwengier) [reply src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs:148]: Will double check, this is probably a holdover

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs
Comment (ryzngard) [reply src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:50]: I think I realized later that we are required to put it on disk for VS (as a comment later in the file)

File: src/Features/Core/Portable/PdbSourceDocument/IPdbSourceDocumentLogger.cs
Comment (ryzngard) [reply src/Features/Core/Portable/PdbSourceDocument/IPdbSourceDocumentLogger.cs:11]: A comment would be nice to help explain that :)

File: src/Features/Core/Portable/PdbSourceDocument/DocumentDebugInfoReader.cs
Comment (ryzngard) [reply src/Features/Core/Portable/PdbSourceDocument/DocumentDebugInfoReader.cs:65]: This was just a curiosity. I think one of the checks is not needed but I'm not sure enough to tell you to change it

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs
Comment (davidwengier) [reply src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs:117]: Will leave this for a follow up 👍

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs
Comment (davidwengier) [reply src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs:148]: Oh, no this is because CreateDocumentInfos takes in a `SourceFileInfo[]`, and `sourceFileInfos` is a `SourceFileInfo?[]`, and we know from the Linq calls that none of them are null. I'll just change CreateDocumentInfos to make it clearer.

**Comment by davidwengier** [comment]
@dotnet/roslyn-infrastructure is there any issues with referencing System.Text.Json here, or the updates to DevDivInsertion.csproj? Would this need to target vs-deps?

**Comment by davidwengier** [comment]
> @dotnet/roslyn-infrastructure is there any issues with referencing System.Text.Json here, or the updates to DevDivInsertion.csproj? Would this need to target vs-deps?

Answer from @dibarbet: As long as the integration tests work and roslyn loads / F5 in public previews it should be fine (I think we're forcing public previews now)

**Comment by davidwengier** [comment]
@jinujoseph for M2 approval. This is the first part of SourceLink, and once merged I'll rebase https://github.com/dotnet/roslyn/pull/58051 which adds the actual debugger dependency to -vs-deps

**Pull request merged**
